xoinvader/collision.py

```diff
@@ -1,49 +1,211 @@
-"""Collision detection."""
-
-
-class SettingsManager(object):
-
-    def __init__(self, settings_map={}):
-        self._smap = settings_map
-
-    def notify(self, owner, settings):
-        self._smap[owner].update(settings)
-
-
-class UpdatableSettings(object):
-
-    def __init__(self, settings_manager, settings={}):
-        self._settings = settings
-        self._settings_manager = settings_manager
-
-    def updateInstanceSettings(self, settings):
-        self._settings.update(settings)
-
-    def notifySettingsManager(self):
-        self._settings_manager.notify(self.__class__.__name__, settings)
-
-
-class CollisionDetector(UpdatableSettings):
-
-    def __init__(self, settings={}):
-        self._settings = settings
-
-
-if __name__ == "__main__":
-    from utils import dotdict
-
-    settings = dotdict({
-        "setting1": True,
-        "setting2": 50,
-        "setting3": "Disabled"
-    })
-
-    settings_manager = SettingsManager({"UpdatableSettings": settings.copy()})
-    print(settings_manager._smap)
-    s1 = UpdatableSettings(settings_manager, settings)
-    print(s1._settings)
-    s1.updateInstanceSettings({"Lolo": 5})
-    print(s1._settings)
-    print(settings_manager._smap)
-    s1.notifySettingsManager()
-    print(settings_manager._smap)
+"""Collision detection system and component module."""
+
+import weakref
+
+from xoinvader.utils import Point
+
+
+class TypePair(object):
+    """Class for hashable unordered string pairs.
+
+    Used as collision dictionary keys, containing pair of collider types. It's
+    set-like, i.e. TypePair(a, b) == TypePair(b, a) and their hashes are equal
+    too.
+
+    :param str a: first collider type
+    :param str b: second collider type
+    """
+
+    def __init__(self, a, b):
+        a, b = sorted([a, b])
+        self._a = a
+        self._b = b
+        self._ab = a + '_' + b
+
+    @property
+    def a(self):
+        return self._a
+
+    @property
+    def b(self):
+        return self._b
+
+    def __eq__(self, other):
+        return self._ab == other._ab
+
+    def __hash__(self):
+        return hash(self._ab)
+
+
+class CollisionManager(object):
+    """Class for collision detection between known components.
+
+    To process collisions, first update the positions of all objects of
+    interest, then call `update` method. It will traverse all registered
+    collisions (between pairs of types) and call appropriate handlers in order
+    they were registered for the two types of colliding objects.
+
+    If you just want to check, if two Colliders collide, call `check_collision`
+    on them.
+    """
+
+    # Marker of solid matter inside collider physics map
+    __solid_matter__ = "#"
+
+    def __init__(self):
+        if Collider.__manager__ is None:
+            def null_manager(_):
+                Collider.__manager__ = None
+            Collider.__manager__ = weakref.ref(self, null_manager)
```

Comment:
It's a "clever hack" to bypass the prohibition of assignment in a lambda. I don't like this to be honest.